starknet_py/net/gateway_client.py

```diff
@@ -0,0 +1,261 @@
+            params=block_identifier,
+        )
+        # TODO should we have better type validation here?
+        return res["amount"]
+
+    async def call_contract(
+        self,
+        invoke_tx: InvokeFunction,
+        block_hash: Optional[Hash] = None,
+        block_number: Optional[int] = None,
+    ) -> List[int]:
+        block_identifier = get_block_identifier(
+            block_hash=block_hash, block_number=block_number
+        )
+
+        res = await self._feeder_gateway_client.post(
+            method_name="call_contract",
+            params=block_identifier,
+            payload=InvokeFunction.Schema().dump(invoke_tx),
+        )
+
+        # TODO should we have better type validation here?
+        return [int(v, 16) for v in res["result"]]
+
+    async def add_transaction(self, tx: StarknetTransaction) -> SentTransaction:
+        res = await self._gateway_client.post(
+            method_name="add_transaction",
+            payload=StarknetTransaction.Schema().dump(obj=tx),
+        )
+        return SentTransactionSchema().load(res, unknown=EXCLUDE)
+
+    async def deploy(
+        self,
+        contract: Union[ContractDefinition, str],
+        constructor_calldata: List[int],
+        salt: Optional[int] = None,
+    ) -> SentTransaction:
+        if isinstance(contract, str):
+            contract = ContractDefinition.loads(contract)
+
+        res = await self.add_transaction(
+            tx=Deploy(
+                contract_address_salt=ContractAddressSalt.get_random_value()
+                if salt is None
+                else salt,
+                contract_definition=contract,
+                constructor_calldata=constructor_calldata,
+                version=0,
+            )
+        )
+
+        if res.code != StarkErrorCode.TRANSACTION_RECEIVED.name:
+            raise TransactionNotReceivedError()
+
+        return res
+
+
+def get_block_identifier(
+    block_hash: Optional[Hash] = None, block_number: Optional[int] = None
+) -> dict:
+    if block_hash is not None and block_number is not None:
+        raise ValueError(
+            "Block_hash and block_number parameters are mutually exclusive."
+        )
+    if block_hash is not None:
+        return {"blockHash": convert_to_felt(block_hash)}
+
+    if block_number is not None:
+        return {"blockNumber": block_number}
+
+    return {}
+
+
+class GatewayHttpClient:
+    def __init__(self, url):
+        self.url = url
+
+    async def call(self, method_name: str, params: Optional[dict] = None) -> dict:
+        address = f"{self.url}/{method_name}"
+
+        async with aiohttp.ClientSession() as session:
+            async with session.get(address, params=params or {}) as request:
+                if request.status != 200:
+                    raise ClientError(
+                        code=str(request.status), message=await request.text()
+                    )
+                return await request.json()
+
+    async def post(
```

Comment:
I think you can use `.request` https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession.trace_config to remove repeated code.